File: gasp/parreno_construct.py
"""Faithful deterministic re-implementation of the Parreno et al. (2008)
block constructive for Container Loading.

KEY DIAGNOSTIC RESULT: this constructive, in a single deterministic pass,
reaches ~88% filling on constrained BR7 with the best-volume criterion --
ABOVE our entire GASP+layout metaheuristic (~87%). It locates the BR
heterogeneity gap precisely: it is in the CONSTRUCTIVE, not the search.
Our EP greedy places one box at a time (~75% on BR7); their block
heuristic places columns/layers of same-type copies, gaining 12-15
points where multiplicity is available. Used as an initial-solution seed
inside GASP (parreno_seed flag), it lifts BR7 from ~87% to ~88.4%, into
the upper band of the literature, close to the top-tier (~91-92%).

The purpose is narrow and diagnostic: reproduce their Section 2
constructive exactly (deterministic, no GRASP randomization, no
improvement phase) and measure it against our own constructive, one
pass each, to locate precisely where the fill gap to the top-tier comes
from. Faithful points, with the two we previously got wrong marked:

  Step 1  choose the maximal space whose nearest vertex-to-container-
          corner distance is lexicographically smallest (volume
          tie-break). The chosen near corner is the anchor.
  Step 2  among all box types fitting the space, enumerate column and
          layer configurations (1..max copies on each axis), evaluate
          each by best-fit (the per-axis gap block-vs-space, sorted,
          lexicographic), tie-break by FEWEST boxes; pick the best.
  *anchor* the block GROWS FROM THE NEAR CORNER inward, not from the
          origin -- this is the fix that makes corner-filling work.
  Step 3  difference process + dominance pruning (shared EMS engine).

Two block-selection objectives are available, as in the paper: best-fit
(criterion ii, default) and best-volume (criterion i).
"""

from __future__ import annotations

from collections import defaultdict
from typing import Dict, List, Tuple

from .ems import _remove_dominated_t, _split_t
from .geometry import Item, Knapsack, Packing, Placement


def _near_corner(s, ks):
    """Return (distance_vector, corner) for the space vertex closest to
    a container corner, lexicographic. corner is one of the 8 vertices
    of the space, expressed as which extreme (min/max) on each axis.

    Optimised: the distance from a space vertex to the nearest container
    corner, per axis, is simply min(coord, dimension - coord); there is
    no need to loop over all 8 container corners. For each of the 8
    space vertices we take, per axis, the smaller of its distance to the
    0-face and to the far face, then sort the triple. This replaces a
    64-iteration loop (8 vertices x 8 corners, each with 3 abs and a
    sort) with 8 cheap evaluations."""
    sx, sy, sz, sx2, sy2, sz2 = s
    W, D, H = ks.W, ks.D, ks.H
    # for each extreme on each axis, distance to the nearer container face
    mx0 = sx if sx < W - sx else W - sx
    mx1 = sx2 if sx2 < W - sx2 else W - sx2
    my0 = sy if sy < D - sy else D - sy
    my1 = sy2 if sy2 < D - sy2 else D - sy2
    mz0 = sz if sz < H - sz else H - sz
    mz1 = sz2 if sz2 < H - sz2 else H - sz2
    # 8 vertices: choose min-face distance per axis depending on which
    # extreme (0=min coord, 1=max coord) the vertex takes on that axis
    verts = (
        ((mx0, my0, mz0), (0, 0, 0)),
        ((mx1, my0, mz0), (1, 0, 0)),
        ((mx0, my1, mz0), (0, 1, 0)),
        ((mx0, my0, mz1), (0, 0, 1)),
        ((mx1, my1, mz0), (1, 1, 0)),
        ((mx1, my0, mz1), (1, 0, 1)),
        ((mx0, my1, mz1), (0, 1, 1)),
        ((mx1, my1, mz1), (1, 1, 1)),
    )
    best = None
    for (a, b, c), sig in verts:
        # sort the 3 distances ascending without a full sort() call
        if a > b: a, b = b, a
        if b > c: b, c = c, b
        if a > b: a, b = b, a
        v = (a, b, c)
        if best is None or v < best[0]:
            best = (v, sig)
    return best


def _place_block(s, sig, w, d, h, nx, ny, nz, members, ncopies):
    """Place ncopies of an oriented box growing from the near corner of
    the space identified by sig (per-axis 0=min,1=max). Returns the list
    of Placements and the block's bounding box (x,y,z,x2,y2,z2)."""
    sx, sy, sz, sx2, sy2, sz2 = s
    bw, bd, bh = nx * w, ny * d, nz * h
    # anchor: if the near corner is at the max side of an axis, the block
    # is flush to that side and grows toward the min; else it starts at min.
    x0 = sx2 - bw if sig[0] == 1 else sx
    y0 = sy2 - bd if sig[1] == 1 else sy
    z0 = sz2 - bh if sig[2] == 1 else sz
    pls = []
    ci = 0
    for ix in range(nx):
        for iy in range(ny):
            for iz in range(nz):
                if ci >= ncopies:
                    break
                it = members[ci]
                pls.append(Placement(it, x0 + ix*w, y0 + iy*d, z0 + iz*h,
                                     w, d, h))
                ci += 1
    return pls, (x0, y0, z0, x0 + bw, y0 + bd, z0 + bh)


def _apply_box(spaces, box):
    x, y, z, x2, y2, z2 = box
    nxt = []
    for sp in spaces:
        if (sp[3] <= x or x2 <= sp[0] or sp[4] <= y or y2 <= sp[1]
                or sp[5] <= z or z2 <= sp[2]):
            nxt.append(sp)
        else:
            nxt.extend(_split_t(sp, x, y, z, x2, y2, z2))
    return _remove_dominated_t(nxt)


def parreno_construct(items: List[Item], ks: Knapsack,
                      allow_rotation: bool = True,
                      objective: str = "bestfit") -> Packing:
    """Deterministic Parreno block constructive. objective in
    {'bestfit', 'bestvol'}."""
    by_type: Dict[tuple, List[Item]] = defaultdict(list)
    for it in items:
        by_type[(it.w, it.d, it.h)].append(it)
    avail = {k: list(v) for k, v in by_type.items()}
    spaces = [(0, 0, 0, ks.W, ks.D, ks.H)]
    placements: List[Placement] = []

    while spaces:
        # Step 1: choose space (min lexicographic near-corner distance,
        # larger volume as tie-break)
        best = None
        for s in spaces:
            dist, sig = _near_corner(s, ks)
            vol = (s[3]-s[0]) * (s[4]-s[1]) * (s[5]-s[2])
            key = (dist, -vol)
            if best is None or key < best[0]:
                best = (key, s, sig)
        _key, s, sig = best
        sx, sy, sz, sx2, sy2, sz2 = s
        fw, fd, fh = sx2 - sx, sy2 - sy, sz2 - sz

        # Step 2: enumerate configurations of every fitting type
        chosen = None
        chosen_score = None
        for tkey, members in avail.items():
            if not members:
                continue
            navail = len(members)
            rep = members[0]
            for (w, d, h) in rep.rotations(allow_rotation, ks.is_3d):
                if w > fw or d > fd or h > fh:
                    continue
                maxx, maxy, maxz = fw // w, fd // d, fh // h
                # enumerate column/layer counts on each axis (full grid is
                # too large; use 1..max per axis but cap product by navail)
                for nx in range(1, maxx + 1):
                    for ny in range(1, maxy + 1):
                        prod_xy = nx * ny
                        if prod_xy > navail:
                            break
                        for nz in range(1, maxz + 1):
                            ncopies = nx * ny * nz
                            if ncopies > navail:
                                break
                            bw, bd, bh = nx*w, ny*d, nz*h
                            if objective == "bestvol":
                                # criterion i: maximise volume, tie fewest
                                score = (-(bw*bd*bh), ncopies)
                            else:
                                # criterion ii: best fit (min gap vector),
                                # tie fewest boxes
                                gap = tuple(sorted((fw-bw, fd-bd, fh-bh)))
                                score = (gap, ncopies)
                            if chosen_score is None or score < chosen_score:
                                chosen_score = score
                                chosen = (tkey, w, d, h, nx, ny, nz, ncopies)
        if chosen is None:
            spaces = [sp for sp in spaces if sp != s]
            continue

        tkey, w, d, h, nx, ny, nz, ncopies = chosen
        members = avail[tkey]
        pls, box = _place_block(s, sig, w, d, h, nx, ny, nz, members, ncopies)
        placements.extend(pls)
        avail[tkey] = members[ncopies:]
        spaces = _apply_box(spaces, box)

    return Packing(ks, placements)

File: gasp_java/src/main/java/gasp/greedy/ParrenoConstruct.java
package gasp.greedy;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Arrays;

import gasp.ems.EMSManager;
import gasp.ems.Space;
import gasp.geometry.Item;
import gasp.geometry.Knapsack;
import gasp.geometry.Packing;
import gasp.geometry.Placement;

public class ParrenoConstruct {

    public static class CornerResult {
        public final int[] dist;
        public final int[] sig;

        public CornerResult(int[] dist, int[] sig) {
            this.dist = dist;
            this.sig = sig;
        }
    }

    public static class BlockResult {
        public final List<Placement> placements;
        public final int[] box; // x, y, z, x2, y2, z2

        public BlockResult(List<Placement> placements, int[] box) {
            this.placements = placements;
            this.box = box;
        }
    }
    
    private static class TypeKey {
        int w, d, h;
        public TypeKey(int w, int d, int h) {
            this.w = w; this.d = d; this.h = h;
        }
        @Override
        public boolean equals(Object o) {
            if (this == o) return true;
            if (o == null || getClass() != o.getClass()) return false;
            TypeKey typeKey = (TypeKey) o;
            return w == typeKey.w && d == typeKey.d && h == typeKey.h;
        }
        @Override
        public int hashCode() {
            int result = w;
            result = 31 * result + d;
            result = 31 * result + h;
            return result;
        }
    }

    public static CornerResult nearCorner(Space s, Knapsack ks) {
        int W = ks.W();
        int D = ks.D();
        int H = ks.H();

        int mx0 = Math.min(s.x, W - s.x);
        int mx1 = Math.min(s.x2, W - s.x2);
        int my0 = Math.min(s.y, D - s.y);
        int my1 = Math.min(s.y2, D - s.y2);
        int mz0 = Math.min(s.z, H - s.z);
        int mz1 = Math.min(s.z2, H - s.z2);

        int[][][] verts = {
            {{mx0, my0, mz0}, {0, 0, 0}},
            {{mx1, my0, mz0}, {1, 0, 0}},
            {{mx0, my1, mz0}, {0, 1, 0}},
            {{mx0, my0, mz1}, {0, 0, 1}},
            {{mx1, my1, mz0}, {1, 1, 0}},
            {{mx1, my0, mz1}, {1, 0, 1}},
            {{mx0, my1, mz1}, {0, 1, 1}},
            {{mx1, my1, mz1}, {1, 1, 1}}
        };

        int[] bestDist = null;
        int[] bestSig = null;

        for (int[][] vert : verts) {
            int a = vert[0][0];
            int b = vert[0][1];
            int c = vert[0][2];
            int[] sig = vert[1];

            // Sort a, b, c ascending
            if (a > b) { int tmp = a; a = b; b = tmp; }
            if (b > c) { int tmp = b; b = c; c = tmp; }
            if (a > b) { int tmp = a; a = b; b = tmp; }

            int[] v = {a, b, c};
            if (bestDist == null || compareIntArrays(v, bestDist) < 0) {
                bestDist = v;
                bestSig = sig;
            }
        }
        return new CornerResult(bestDist, bestSig);
    }

    private static int compareIntArrays(int[] a, int[] b) {
        for (int i = 0; i < Math.min(a.length, b.length); i++) {
            if (a[i] != b[i]) {
                return Integer.compare(a[i], b[i]);
            }
        }
        return Integer.compare(a.length, b.length);
    }

    public static BlockResult placeBlock(Space s, int[] sig, int w, int d, int h, int nx, int ny, int nz, List<Item> members, int ncopies) {
        int bw = nx * w;
        int bd = ny * d;
        int bh = nz * h;

        int x0 = (sig[0] == 1) ? s.x2 - bw : s.x;
        int y0 = (sig[1] == 1) ? s.y2 - bd : s.y;
        int z0 = (sig[2] == 1) ? s.z2 - bh : s.z;

        List<Placement> pls = new ArrayList<>();
        int ci = 0;
        for (int ix = 0; ix < nx; ix++) {
            for (int iy = 0; iy < ny; iy++) {
                for (int iz = 0; iz < nz; iz++) {
                    if (ci >= ncopies) {
                        break;
                    }
                    Item it = members.get(ci);
                    pls.add(new Placement(it, x0 + ix * w, y0 + iy * d, z0 + iz * h, w, d, h));
                    ci++;
                }
            }
        }
        int[] box = {x0, y0, z0, x0 + bw, y0 + bd, z0 + bh};
        return new BlockResult(pls, box);
    }

    public static List<Space> applyBox(List<Space> spaces, int[] box) {
        return EMSManager.differenceProcess(spaces, box, 1);
    }

    public static Packing parrenoConstruct(List<Item> items, Knapsack ks, boolean allowRotation, String objective) {
        Map<TypeKey, List<Item>> avail = new java.util.LinkedHashMap<>();
        for (Item it : items) {
            TypeKey key = new TypeKey(it.w(), it.d(), it.h());
            avail.computeIfAbsent(key, k -> new ArrayList<>()).add(it);
        }

        List<Space> spaces = new ArrayList<>();
        spaces.add(new Space(0, 0, 0, ks.W(), ks.D(), ks.H()));
        List<Placement> placements = new ArrayList<>();

        while (!spaces.isEmpty()) {
            CornerResult bestCorner = null;
            Space bestSpace = null;
            long bestVol = 0;
            
            for (Space s : spaces) {
                CornerResult cr = nearCorner(s, ks);
                long vol = s.getVolume();
                
                if (bestCorner == null) {
                    bestCorner = cr;
                    bestSpace = s;
                    bestVol = vol;
                } else {
                    int cmp = compareIntArrays(cr.dist, bestCorner.dist);
                    if (cmp < 0 || (cmp == 0 && vol > bestVol)) {
                        bestCorner = cr;
                        bestSpace = s;
                        bestVol = vol;
                    }
                }
            }
            
            Space s = bestSpace;
            int fw = s.getW();
            int fd = s.getD();
            int fh = s.getH();

            Object[] chosenScore = null;
            TypeKey chosenKey = null;
            int[] chosenParams = null; // {w, d, h, nx, ny, nz, ncopies}

            for (Map.Entry<TypeKey, List<Item>> entry : avail.entrySet()) {
                TypeKey tkey = entry.getKey();
                List<Item> members = entry.getValue();
                if (members.isEmpty()) continue;
                int navail = members.size();
                Item rep = members.get(0);

                for (int[] rot : rep.rotations(allowRotation, ks.is3D())) {
                    int w = rot[0], d = rot[1], h = rot[2];
                    if (w > fw || d > fd || h > fh) continue;

                    int maxx = fw / w;
                    int maxy = fd / d;
                    int maxz = fh / h;

                    for (int nx = 1; nx <= maxx; nx++) {
                        for (int ny = 1; ny <= maxy; ny++) {
                            if (nx * ny > navail) break;
                            for (int nz = 1; nz <= maxz; nz++) {
                                int ncopies = nx * ny * nz;
                                if (ncopies > navail) break;

                                int bw = nx * w, bd = ny * d, bh = nz * h;
                                Object[] score;
                                if ("bestvol".equals(objective)) {
                                    score = new Object[]{(long) -(bw * bd * bh), ncopies};
                                } else {
                                    int[] gap = {fw - bw, fd - bd, fh - bh};
                                    Arrays.sort(gap);
                                    score = new Object[]{gap, ncopies};
                                }

                                if (chosenScore == null || compareScores(score, chosenScore, objective) < 0) {
                                    chosenScore = score;
                                    chosenKey = tkey;
                                    chosenParams = new int[]{w, d, h, nx, ny, nz, ncopies};
                                }
                            }
                        }
                    }
                }
            }

            if (chosenParams == null) {
                spaces.remove(s);
                continue;
            }

            int w = chosenParams[0], d = chosenParams[1], h = chosenParams[2];
            int nx = chosenParams[3], ny = chosenParams[4], nz = chosenParams[5], ncopies = chosenParams[6];
            List<Item> members = avail.get(chosenKey);

            BlockResult br = placeBlock(s, bestCorner.sig, w, d, h, nx, ny, nz, members, ncopies);
            placements.addAll(br.placements);
            avail.put(chosenKey, members.subList(ncopies, members.size()));
            spaces = applyBox(spaces, br.box);
        }

        return new Packing(ks, placements);
    }

    private static int compareScores(Object[] a, Object[] b, String objective) {
        if ("bestvol".equals(objective)) {
            long aVol = (long) a[0];
            long bVol = (long) b[0];
            if (aVol != bVol) return Long.compare(aVol, bVol);
        } else {
            int[] aGap = (int[]) a[0];
            int[] bGap = (int[]) b[0];
            int cmp = compareIntArrays(aGap, bGap);
            if (cmp != 0) return cmp;
        }
        int aCopies = (int) a[1];
        int bCopies = (int) b[1];
        return Integer.compare(aCopies, bCopies);
    }
}
